### PR TITLE
PS-3834: Implement innodb log scrubbing

### PR DESCRIPTION
**Pull request by dutow**
Based on the same feature in MariaDB

### CODE REVIEW

File: storage/innobase/include/log0log.h [hunk 1]
```diff
@@ -238,6 +238,10 @@ log_make_checkpoint_at(
 	lsn_t			lsn,
 	bool			write_always);
 
+void
+log_ensure_scrubbing_thread(void);
```
Comment (laurynas-biveinis) [inline storage/innobase/include/log0log.h:242]: extern "C" ?

File: storage/innobase/log/log0log.cc [hunk 2]
```diff
@@ -124,6 +128,15 @@ the previous */
...
 #define LOG_UNLOCK_FLUSH_LOCK		2
 
+/** Event to wake up log_scrub_thread */
+os_event_t	log_scrub_event;
+/** Whether log_scrub_thread is active */
+bool		log_scrub_thread_active;
+
+extern "C"
+os_thread_ret_t
+DECLARE_THREAD(log_scrub_thread)(void*);
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:138]: Redundant declaration

File: storage/innobase/log/log0log.cc [hunk 3]
```diff
@@ -2291,12 +2304,17 @@ logs_empty_and_mark_files_at_shutdown(void)
...
 	log_mutex_enter();
 	const ulint	n_write	= log_sys->n_pending_checkpoint_writes;
 	const ulint	n_flush	= log_sys->n_pending_flushes;
 	log_mutex_exit();
 
-	if (n_write != 0 || n_flush != 0) {
+	if (log_scrub_thread_active || n_write != 0 || n_flush != 0) {
 		if (srv_print_verbose_log && count > 600) {
 			ib::info() << "Pending checkpoint_writes: " << n_write
 				<< ". Pending log flush writes: " << n_flush;
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:2320]: Diagnose scrubber running too

File: storage/innobase/log/log0log.cc [hunk 4]
```diff
@@ -2646,4 +2670,113 @@ log_mem_free(void)
 		log_sys = NULL;
 	}
 }
+
+void
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:2674]: Header comment

File: storage/innobase/include/log0log.h [hunk 5]
```diff
@@ -238,6 +238,10 @@ log_make_checkpoint_at(
 	lsn_t			lsn,
 	bool			write_always);
 
+void
```
Comment (laurynas-biveinis) [inline storage/innobase/include/log0log.h:241]: Header comment

File: storage/innobase/include/srv0srv.h [hunk 6]
```diff
@@ -1103,6 +1109,8 @@ struct export_var_t{
 
 	fragmentation_stats_t innodb_fragmentation_stats;/*!< Fragmentation
 						statistics */
+
+	int64_t innodb_scrub_log;
```
Comment (laurynas-biveinis) [inline storage/innobase/include/srv0srv.h:1113]: Comment (it's genuinely not clear, what's counted here?)

File: storage/innobase/log/log0log.cc [hunk 7]
```diff
@@ -2646,4 +2670,113 @@ log_mem_free(void)
...
+log_ensure_scrubbing_thread(void)
+{
+	log_scrub_thread_active = srv_scrub_log;
+	if (log_scrub_thread_active) {
+		log_scrub_event = os_event_create("log_scrub_event");
+		os_thread_create(log_scrub_thread, NULL, NULL);
+	}
+}
+
+/******************************************************//**
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:2684]: New InnoDB doxygen style comment

File: storage/innobase/log/log0log.cc [hunk 8]
```diff
@@ -2646,4 +2670,113 @@ log_mem_free(void)
...
+Pads the current log block full with dummy log records. Used in producing
+consistent archived log files and scrubbing redo log. */
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:2686]: There is no "consistent archived log files" in 5.7 innodb

File: storage/innobase/log/log0log.cc [hunk 9]
```diff
@@ -2646,4 +2670,113 @@ log_mem_free(void)
...
+log_pad_current_log_block(void)
+/*===========================*/
+{
+	byte		b		= MLOG_DUMMY_RECORD;
+	ulint		pad_length;
+	ulint		i;
+	lsn_t		lsn;
+
+	ut_ad(!recv_no_log_write);
+	/* We retrieve lsn only because otherwise gcc crashed on HP-UX */
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:2698]: Let's not

File: storage/innobase/log/log0log.cc [hunk 10]
```diff
@@ -2646,4 +2670,113 @@ log_mem_free(void)
...
+void
+log_pad_current_log_block(void)
+/*===========================*/
+{
+	byte		b		= MLOG_DUMMY_RECORD;
+	ulint		pad_length;
+	ulint		i;
+	lsn_t		lsn;
+
+	ut_ad(!recv_no_log_write);
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:2697]: Assert not read only mode too

File: storage/innobase/log/log0log.cc [hunk 11]
```diff
@@ -2646,4 +2670,113 @@ log_mem_free(void)
...
+
+	lsn = log_sys->lsn;
+
+	log_close();
+
+	ut_a(lsn % OS_FILE_LOG_BLOCK_SIZE == LOG_BLOCK_HDR_SIZE);
+}
+
+
+/*****************************************************************//*
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:2727]: New InnoDB doxygen style comments

File: storage/innobase/log/log0log.cc [hunk 12]
```diff
@@ -2646,4 +2670,113 @@ log_mem_free(void)
...
+
+	next_lbn_to_pad = log_block_convert_lsn_to_no(log_sys->lsn);
+
+  log_mutex_exit();
+}
+
+/* log scrubbing speed, in bytes/sec */
+ulonglong innodb_scrub_log_speed;
+
+/*****************************************************************//**
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:2751]: Likewise

File: storage/innobase/log/log0log.cc [hunk 13]
```diff
@@ -2646,4 +2670,113 @@ log_mem_free(void)
...
+This is the main thread for log scrub. It waits for an event and
+when waked up fills current log block with dummy records and
+sleeps again.
+@return this function does not return, it calls os_thread_exit() */
+extern "C"
+os_thread_ret_t
+DECLARE_THREAD(log_scrub_thread)(void*)
+{
+	ut_ad(!srv_read_only_mode);
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:2760]: This thread is invisible for performance schema

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
This is not mergeable PR until redo log itself is encrypted in the first place. No point in addressing review comments until then neither (as bulk of them are about tests, which will need it)

Please add feature tests (enable that option, and issue write workload, both at full speed and throttled).
Add a small testcase that innodb_read_only is handled correctly

**Comment by laurynas-biveinis** [comment]
Once it can be actually tested, run the whole testsuite with scrubbing enabled, collect any meaningful failures to new regression testcases

**Comment by dutow** [comment]
Closing this PR, as it's now part of the "everything related to scrubbing" PR:
https://github.com/robgolebiowski/percona-server/pull/3/files

**Pull request closed without merging**